### PR TITLE
Refactors handlers to reduce complexity & side-effects

### LINKED ISSUE #241: refactor contentful ns
from @tylerperkins PR#155

"We shouldn't be complecting the update of the db with hitting Contentful. There is a lot of unnecessary complexity, like storing route parameters in the db."

"each of the activity, branch, skill, and platform, cases should have their own handlers. For the `filtered-activities` view and `activity-thumbnail` component, they both do some calculation to get the map at `[:app :route-params]`, then get its first key and check to see if it is `:platform`. It would be far better to simply assign a boolean value at, say, `[:app :viewing-platform?]`. That would be specific, efficient, and would make future misuse unlikely. Since this boolean value could change with each request, it should simply be passed as a parameter from the router to any handler that needs it."

### PR DESCRIPTION
**Pull request by eemshi**
v.2

closes https://github.com/codefordenver/owlet/issues/241

TODO:
- [x] get rid of `:process-activity-metadata` dispatch in `:get-library-content-from-contentful`
- [x] stop storing route-params
- [x] resolve timing of `:get-library-content-from-contentful` and `:filter-activities-by-search-term` (see `routes.cljs`)
- [ ] handle `set-path` in `:filter-activities-by-search-term` as an effect? **[deferred]**
- [x] consolidate `:set-activities-by-branch-in-view` with `:filter-activities-by-search-term`
- [x] update error message for search not found
- [ ] revisit `:set-active-document-title!` **[deferred]**

### CODE REVIEW

**Comment by eemshi** [comment]
_migrated comment from @tylerperkins_ 

A good start; much cleaner! Some notes:
  - Keywords are functions like any other, so `#(:fields %)` can be replaced by `:fields`.
  - `(-> db (assoc ...))` can shortened to `(assoc db ...)`
  - It occurs to me that the multiple dispatches performed by the route handlers in routes.cljs are problematic: 

    First, if they depend on each other, things could get confusing. Second, they don't really represent _events_ from the outside world. I think that for each route, you really mean to do just _one_ dispatch that composes a few functions together, and that's exactly what interceptors are for. How about making separate event handlers :show-branches, :show-skill, :show-platform, etc. Then within each handler, employ interceptors as a pipeline given in the handler. You've probably already thought of this for "handle `set-path` in `:filter-activities-by-search-term` as a coeffect", but I think this technique is worth considering for most of those multiple dispatches.

    For example, :show-branches could designate `[get-library-content-from-contentful set-active-view]` as its interceptor pipeline, where these are all pure functions just inputing and outputting a context value. Here, `get-library-content-from-contentful` has to asynchronously issue a GET request, and that is an effect. So it would be registered like this
    ```
    (re-frame.core/->interceptor
          :id      :get-library-content-from-contentful
          :before  (fn [context]
                     (assoc-in context [:effects :get-library] [])
    ```
    Here, :get-library would be registered as an effect using `reg-fx`. The function associated there would do just what the current :get-library-content-from-contentful _event_ handler does. But now, we have made all our interceptors and handlers pure, except for the effect handlers registered with `reg-fx`. That's the best practice with re-frame, as I understand it.

**Comment by eemshi** [comment]
_migrated comment from @tylerperkins_ 

I was thinking that maybe I oversold the interceptor idea. It would be perfectly fine to refer to effect handlers (defined using `reg-fx`) in the map each of your event handlers return. No interceptors required; they might just streamline things a bit in some cases. The important point is that the handler functions (and interceptors) be pure, with effects like API requests or direct DOM manipulation living in effect handlers. "_**Derived Values, Flowing**_"!

**Comment by eemshi** [comment]
~@tylerperkins I think maybe an interceptor _is_ needed here because of the timing issue. If the API request in `:get-library-content-from-contentful` succeeds, it then dispatches `:get-library-content-from-contentful-successful`, which populates `app-db` with the data we need in order to display activities. Since there is no guaranteed [order of effects](https://github.com/Day8/re-frame/blob/master/docs/Effects.md#order-of-effects), maybe an interceptor will give us the extra milliseconds needed to finish `:get-library-content-from-contentful-successful` before doing the main handler. otherwise i might have to use `:dispatch-later` in the main handler~

**Comment by eemshi** [comment]
nevermind @tylerperkins, i ended up doing this a different way - let me know what you think

File: src/cljs/owlet_ui/events/contentful.cljs [hunk 1]
```diff
@@ -70,24 +69,41 @@
...
     {:dispatch-n (list [:set-active-view :filtered-activities-view "branch"]
-                       [:filter-activities-by-search-term (second route-param)])}))
+                       [:filter-activities-by-search-term route-param]
+                       [:set-active-document-title! route-param])}))
+
+(rf/reg-event-fx
+  :show-activity
+  (fn [_ [_ route-param]]
+    {:dispatch-n (list [:set-active-view :activity-view]
+                       [:set-activity-in-view route-param])}))
```
Comment (eemshi) [inline src/cljs/owlet_ui/events/contentful.cljs:106]: @dviramontes this is now the only place `:set-activity-in-view` is being used. `[:filter-activities-by-search-term route-param]` also works here - do you see any advantage in using `:set-activity-in-view` instead? i was thinking speed maybe

Comment (tylerperkins) [reply src/cljs/owlet_ui/events/contentful.cljs:106]: Definitely heading in the right direction! You can see another example of this event-callback technique in my `:auth0-authenticated` effect, which gets fired by a call back invoked by the Auth0Lock JS object. It in turn invokes effect `:firebase-sign-in` with an argument `:firebase-sign-in-failed`. That event handler is in turn invoked if Auth0 couldn't sign into Firebase. So everything proceeds in order: function --> effect --> effect --> event!

Re. use of the `:dispatch-n` or `:dispatch` keys: Yes, this makes your handler pure, but the result is the same, doing a bunch of dispatches that could be impure. If the handler you're dispatching to is pure, then don't dispatch. Just call it as a function. We need to be rewriting them as follows:
```
(rf/reg-event-db
  :set-activity-in-view
  (fn [db [_ activity-id all-activities]]
    ... ))

;; BECOMES

(defn set-activity-in-view
  [db [_ activity-id all-activities]]
  ... )
(rf/reg-event-db :set-activity-in-view set-activity-in-view)
```
If, on the other hand, the dispatched-to event handler is impure, then it should be rewritten as an _effect_ handler, i.e., in a call to `reg-fx`. In rare cases like `:set-active-view`, the DOM-manipulating effect probably belongs in a view instead. The DOM-accessing code could be rewritten as JavaScript in a `<script>` tag. The tag, along with the rest of the page, would be built by our view function. That way, we just render the page as normal and the effect stays out there in the impure DOM world, not in our handlers.

Think of events (along with coeffects) as the way _the outside world_ talks to our app, not the way the app talks to itself. Our app talks to the outside world through views and effects. Everything else should be pure!

Comment (eemshi) [reply src/cljs/owlet_ui/events/contentful.cljs:106]: makes sense! i'd like to merge this branch and open a new one for refactoring `:set-active-document-title!` and `:filter-activities-by-search-term`. could i get a review for merging @tylerperkins @dviramontes ?

File: src/cljs/owlet_ui/events/contentful.cljs [hunk 2]
```diff
@@ -69,14 +60,105 @@
...
-          (cond platform (rf/dispatch [:filter-activities-by-search-term platform])
-                skill    (rf/dispatch [:filter-activities-by-search-term skill])
-                activity (rf/dispatch [:set-activity-in-view activity all-activities])
-                branch   (let [activities-by-branch-in-view ((keyword branch) activities-by-branch)]
-                          (rf/dispatch [:set-activities-by-branch-in-view branch activities-by-branch-in-view])))))
-      (assoc db :activity-branches branches
-                :skills skills
-                :activities-by-branch activities-by-branch
-                :activity-titles activity-titles))))
+      {:db (assoc db
```
Comment (dviramontes) [inline src/cljs/owlet_ui/events/contentful.cljs:63]: 👏

File: src/cljs/owlet_ui/events/contentful.cljs (hunk 1 above)
Comment (dviramontes) [reply src/cljs/owlet_ui/events/contentful.cljs:106]: 💯

File: src/cljs/owlet_ui/routes.cljs [hunk 3]
```diff
@@ -29,26 +29,19 @@
             (rf/dispatch [:set-active-view :unsubscribe-view]))
 
   (defroute "/branches" []
-            (rf/dispatch [:get-library-content-from-contentful])
-            (rf/dispatch [:set-active-view :branches-view])
-            (rf/dispatch [:set-active-document-title! "Branches"]))
+            (rf/dispatch [:get-content-from-contentful :show-branches]))
```
Comment (dviramontes) [inline src/cljs/owlet_ui/routes.cljs:32]: 🚀

File: src/cljs/owlet_ui/app.cljs [hunk 4]
```diff
@@ -22,7 +22,7 @@
 (defmulti views identity)
 (defmethod views :welcome-view [] [welcome-view])
 (defmethod views :filtered-activities-view []
-  [filtered-activities-view (first (keys @(rf/subscribe [:route-params])))])
+  [filtered-activities-view])
```
Comment (dviramontes) [inline src/cljs/owlet_ui/app.cljs:25]: 👍

**Review by dviramontes — APPROVED** [review]
🥇 thanks @eemshi! This looks great!

**Pull request merged**